src/gromacs/nblib/topology.cpp

```diff
@@ -106,11 +106,39 @@ std::vector<real> TopologyBuilder::extractQuantity(Extractor extractor)
     return ret;
 }
 
+std::vector<real> TopologyBuilder::extractCharge()
+{
+    auto &moleculesList = molecules_;
+
+    //! returned object
+    std::vector<real> ret;
+    ret.reserve(numAtoms_);
+
+    for (auto &molNumberTuple : moleculesList)
+    {
+        Molecule &molecule = std::get<0>(molNumberTuple);
+        size_t numMols = std::get<1>(molNumberTuple);
+
+        for (size_t i = 0; i < numMols; ++i)
+        {
+            for (auto &atomTuple : molecule.atoms_)
+            {
+                std::string atomTypeName = std::get<1>(atomTuple);
+
+                real charge = std::get<1>(molecule.atomTypes_[atomTypeName]);
+                ret.push_back(charge);
+            }
+        }
+    }
+
+    return ret;
+}
+
```

Comment:
I'm pretty sure we can avoid duplicating the function and slightly adapt the lambda to extract the relevant tuple member. I was going to try this, but I can't compile this branch.